vulnerabilities/pipelines/remove_duplicate_advisories.py

```diff
@@ -0,0 +1,83 @@
+# VulnerableCode is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/aboutcode-org/vulnerablecode for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+import logging
+from itertools import groupby
+
+from aboutcode.pipeline import LoopProgress
+from django.db.models import Count
+from django.db.models import Q
+
+from vulnerabilities.models import Advisory
+from vulnerabilities.pipelines import VulnerableCodePipeline
+from vulnerabilities.utils import compute_content_id
+
+
+class RemoveDuplicateAdvisoriesPipeline(VulnerableCodePipeline):
+    """Pipeline to remove duplicate advisories based on their content."""
+
+    pipeline_id = "remove_duplicate_advisories"
+
+    @classmethod
+    def steps(cls):
+        return (
+            cls.recompute_content_ids,
+            cls.remove_duplicates,
+        )
+
+    def remove_duplicates(self):
+        """
+        Find advisories with the same content and keep only the latest one.
+        """
+
+        duplicated_advisories = groupby(
+            Advisory.objects.order_by("unique_content_id").all().paginated(),
+            key=lambda x: x.unique_content_id,
+        )
+        progress = LoopProgress(total_iterations=Advisory.objects.count(), logger=self.log)
+        for _content_id, advisories in progress.iter(duplicated_advisories):
+            advisories = list(advisories)
+            self.log(
+                f"Removing duplicates for content ID {_content_id} {len(advisories)}",
+                level=logging.INFO,
+            )
+            oldest = min(advisories, key=lambda x: x.date_imported)
+            try:
+                advisory_ids = []
+                for adv in advisories:
+                    if adv.id != oldest.id:
+                        advisory_ids.append(adv.id)
+                Advisory.objects.filter(id__in=advisory_ids).delete()
+            except Exception as e:
+                self.log(f"Error deleting advisories: {e}", level=logging.ERROR)
+
+            self.log(
+                f"Kept advisory {oldest.id} and removed "
+                f"{len(list(advisories)) - 1} duplicates for content ID {_content_id}",
+                level=logging.INFO,
+            )
+
+    def recompute_content_ids(self):
+        """
+        Recompute content IDs for all advisories.
+        """
+
+        advisories = []
+
+        progress = LoopProgress(
+            total_iterations=Advisory.objects.count(),
+            progress_step=1,
+            logger=self.log,
+        )
+
+        for advisory in progress.iter(Advisory.objects.all().paginated()):
+            advisory.unique_content_id = compute_content_id(advisory)
+            advisories.append(advisory)
+
+        Advisory.objects.bulk_update(advisories, ["unique_content_id"], batch_size=1000)
```

Comment:
Keeping all that `advisories` in memory is going to be very expensive!
Assuming 1 Advisory object ≈ 1000 bytes.
That would mean for 118 million advisories: (118 * 10^6) * 10^3 bytes ≈ 118 GB of memory!.

We should not keep all these advisories in memory. Instead we should bulk update as soon as we reach the batch size.